web3/main.py

```diff
@@ -237,13 +243,13 @@ def toChecksumAddress(value: Union[AnyAddress, str, bytes]) -> ChecksumAddress:
 
     def __init__(
         self,
-        provider: Optional[BaseProvider] = None,
+        provider: Optional[Union[BaseProvider, AsyncBaseProvider]] = None,
         middlewares: Optional[Sequence[Any]] = None,
         modules: Optional[Dict[str, Union[Type[Module], Sequence[Any]]]] = None,
         external_modules: Optional[
             Dict[str, Union[Type[Module], Sequence[Any]]]
         ] = None,
-        ens: ENS = cast(ENS, empty),
+        ens: Optional[Union[ENS, AsyncENS]] = None,
```

Comment:
Should this still be `empty`? or does it get set somewhere else?

Reply:
Thanks for catching this string of typing. This is the main part I struggled to make the linting work then focused on something else and I admit this isn't ideal at all.

I switched a few times between `empty` and `None` but `None` can sometimes get into recursion... again not really sure what the right setup is here. In the end I went with `cast(ENS, empty)` and we can return `AsyncENS` when it's appropriate (when the eth module is async). Thoughts?